internal/jobcompleter/job_completer.go

```diff
@@ -56,125 +63,299 @@ type InlineJobCompleter struct {
+	c.errGroup.Go(func() error {
+		job, err := withRetries(ctx, &c.BaseService, func(ctx context.Context) (*rivertype.JobRow, error) {
+			return c.exec.JobSetStateIfRunning(ctx, params)
+		})
 		if err != nil {
 			return err
 		}
 
 		stats.CompleteDuration = c.TimeNowUTC().Sub(start)
+		c.sendJobToSubscription(job, stats)
+
+		return nil
+	})
+	return nil
+}
+
+func (c *AsyncCompleter) Start(ctx context.Context) error { return nil }
+
+func (c *AsyncCompleter) Stop() {
+	if err := c.errGroup.Wait(); err != nil {
+		c.Logger.Error("Error waiting on async completer: %s", err)
+	}
+}
+
+type batchCompleterSetState struct {
+	Params    *riverdriver.JobSetStateIfRunningParams
+	Stats     *jobstats.JobStatistics
+	WaitingAt time.Time // went job was submitted for completion
+}
+
+// BatchCompleter uses a debounced channel to accumulate incoming completions
+// and every so often complete many of them as a single efficient batch. To
+// minimize the amount of driver surface area we need, the batching is only
+// performed for jobs being changed to a `completed` state, which we expect to
+// the vast common case under normal operation. The completer embeds an
+// AsyncCompleter to perform other non-`completed` state completions.
+type BatchCompleter struct {
+	baseservice.BaseService
+	startstop.BaseStartStop
+	withSubscribe
+
+	asyncCompleter   *AsyncCompleter // used for non-complete completions
+	debounceChan     *chanutil.DebouncedChan
+	exec             PartialExecutor
+	ready            chan struct{}
+	setStateParams   map[int64]*batchCompleterSetState
+	setStateParamsMu sync.Mutex
+}
+
+func NewBatchCompleter(archetype *baseservice.Archetype, exec PartialExecutor) *BatchCompleter {
+	return baseservice.Init(archetype, &BatchCompleter{
+		asyncCompleter: NewAsyncCompleter(archetype, exec),
+		exec:           exec,
+		setStateParams: make(map[int64]*batchCompleterSetState),
+	})
+}
+
+func (c *BatchCompleter) Start(ctx context.Context) error {
+	stopCtx, shouldStart, stopped := c.StartInit(ctx)
+	if !shouldStart {
+		return nil
+	}
+
+	c.ready = make(chan struct{})
+
+	go func() {
+		// This defer should come first so that it's last out, thereby avoiding
+		// races.
+		defer close(stopped)
+
+		c.Logger.InfoContext(ctx, c.Name+": Run loop started")
+		defer c.Logger.InfoContext(ctx, c.Name+": Run loop stopped")
 
-		func() {
-			c.subscribeFuncMu.Lock()
-			defer c.subscribeFuncMu.Unlock()
+		c.debounceChan = chanutil.NewDebouncedChan(stopCtx, 100*time.Millisecond)
 
-			if c.subscribeFunc != nil {
-				c.subscribeFunc(CompleterJobUpdated{Job: job, JobStats: stats})
+		close(c.ready)
+
+		for {
+			select {
+			case <-stopCtx.Done():
+				// Try to insert last batch before leaving. Note we use the
+				// original context so operations aren't immediately cancelled.
+				if err := c.handleBatch(ctx); err != nil {
+					c.Logger.Error(c.Name+": Error completing batch", "err", err)
+				}
+
+				return
+			case <-c.debounceChan.C():
```

Comment:
Doesn't this `DebouncedChan` type fire on the leading edge? So that means we will always submit a batch of 1 before then waiting to collect additional jobs? Doesn't seem ideal if so.

Reply:
True.

So this one is a bit of a story -- I modified `DebounceChan` to take a boolean in its initializer for `sendLeaading` for use in cases like this, and added a test.

_But_, then I modified the loop with the job threshold mentioned above, and since it's possible for the loop to skip completing a batch if not enough jobs have accumulated yet, the use of `DebounceChan` doesn't make sense anymore. Instead, it should just be a normal `time.Ticker`.

Changed it and it works fine. I left the `sendLeading` option in `DebounceChan` mentioned above since I'd already written a test case for it, but I guess this could be reverted.